src/protect/utils.php

```diff
@@ -0,0 +1,36 @@
+<?php
+function is_url_whitelisted (string $redirect_url, array $whitelisted_origins) {
+    // Required to check if url is valid, otherwise parse_url may not work
+    if (!str_starts_with($redirect_url, "http") &&
+        filter_var($redirect_url, FILTER_VALIDATE_URL))
+        return false;
+
+    $parsed_url = parse_url($redirect_url);
+    // To be absolutely sure there is no error when accessing keys later
+    if (!array_key_exists("scheme", $parsed_url) ||
+        !array_key_exists("host", $parsed_url))
+        return false;
+
+    for ($i = 0 ; $i <  count($whitelisted_origins) ; ++$i) {
+        $parsed_origin = parse_url($whitelisted_origins[$i]);
```

Comment:
I would have put a `foreach` loop here instead
```suggestion
    foreach ($whitelisted_origins as $origin) {
        $parsed_origin = parse_url($origin);
```